### ISSUE TITLE
Tracking Issue for `core::future::join!`

### ISSUE THREAD

**Issue by ibraheemdev** [rust-lang/rust#91642]
Feature gate: `#![feature(future_join)]`

This is a tracking issue for `core::future::join!`.

```rust
let (a, b) = join!(a(), b());
```

### Public API

```rust
// core::future

pub macro join {
    ( $($fut:expr),* $(,)?) => {
        // ...
    }
}
```

### Steps / History

- [ ] Implementation: #...
- [ ] Final comment period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

**Comment by yoshuawuyts** [comment]
The @rust-lang/wg-async-foundations group was not notified of https://github.com/rust-lang/rust/pull/91645 prior to merging. Tagging here to ensure the right people get the notification that this was merged into unstable.

**Comment by yoshuawuyts** [comment]
I don't think that `#![feature(future_join)]` should have been merged as-is, and instead needs to be part of a coherent design implementing async concurrency for Rust. The topic of async concurrency is a fundamental part of Rust's future, and nuanced enough that it should be going through the RFC process [^prior].

[^prior]: See [Swift SE-0304: Structured Concurrency](https://github.com/apple/swift-evolution/blob/main/proposals/0304-structured-concurrency.md) for an example of what scope I expect a Rust RFC on async concurrency to cover.

Personally I've found the abstractions provided in `futures-rs` (e.g. `join!`, `join_all!`, `select! {}` et. al.) to be severely lacking. For that reason we never adopted these in [`async_std::future`](https://docs.rs/async-std/latest/async_std/future/index.html), and have instead been researching alternatives ([1], [2], [3], [4], [5], [6]) for the past three years.

[1]: https://blog.yoshuawuyts.com/futures-concurrency/
[2]: https://blog.yoshuawuyts.com/streams-concurrency/
[3]: https://blog.yoshuawuyts.com/parallel-stream/
[4]: https://blog.yoshuawuyts.com/future-join-and-const-eval/
[5]: https://blog.yoshuawuyts.com/futures-concurrency-2/
[6]: https://docs.rs/futures-concurrency/1.0.0/futures_concurrency/

Because of the size of the topic, I would like to see a project group started as part of the [async foundations WG](https://rust-lang.github.io/wg-async-foundations/vision/roadmap.html#overview), tasked with describing a coherent model for async concurrency in Rust. The goal would be to produce an RFC + implementation of async concurrency adapters for the stdlib, which can then be implemented as part of the stdlib.

Which brings us back to what to do with `#![feature(future_join)]`. I see two plausible options here:
1. We change the feature gate to `#[feature(future_concurrency)]`, and file a blocking concern mentioning this requires an RFC.
2. We file a PR removing the feature from the stdlib, close the tracking issue, and start fresh with a group under the async foundations WG.

I want to ensure we don't accidentally stabilize this issue without having considered the much wider topic of creating a coherent model and corresponding interfaces for async concurrency in Rust.

**Comment by carllerche** [comment]
I agree w/ @yoshuawuyts and don't have much else to add. Tokio added `join!` mostly because it was an established pattern. If I were to do it again, I would not have included `join!`. After onboarding quite a few new devs to async rust, I think `join!` complicates the learning experience without adding **much** over spawning the sub-tasks. There are alternative strategies that could be explored over `join!`.

**Comment by eholk** [comment]
> After onboarding quite a few new devs to async rust, I think join! complicates the learning experience without adding much over spawning the sub-tasks.

@carllerche - I was wondering if you could provide more details about your experience here? What did new devs find complicated about `join!`?

**Comment by cryptoquick** [comment]
Is there a better way to accomplish parallel await than with this feature? My use-case is basically the example here:
https://doc.rust-lang.org/std/future/macro.join.html

**Comment by yoshuawuyts** [comment]
@cryptoquick the stdlib doesn't provide any other APIs for concurrenly `.await`ing futures right now, no. You can look at the `futures`, `async-std`, `tokio`, and `futures-lite` ecosystem crates for various variations on the `join!` functionality provided here.

Additionally I've been experimenting with alternative static async concurrency designs and writing about them in my [blog](https://blog.yoshuawuyts.com/futures-concurrency-3/), with the intent of arriving at a design we can put forward for stabilization. There's a crate published now too: [`futures-concurrency`](https://docs.rs/futures-concurrency/latest/futures_concurrency/), which implements some of these designs. Note that right now this only covers _static_ fixed-sized concurrency such as `join!`, and not _dynamic_ variable-sized concurrency which is provided by abstractions such as `FuturesUnordered` and `StreamMap`.

As I mentioned in https://github.com/rust-lang/rust/issues/91642#issuecomment-992773288 adding async concurrency APIs to the stdlib is a large enough topic that it should go through an RFC. Until we've done that, I think it's unlikely we'll be seeing any concurrency functionality stabilized in the stdlib.

I hope that answers your question!

**Comment by discreaminant2809** [comment]
Should it accept `IntoFuture` instead of `Future`?

**Comment by RalfJung** [comment]
Looks like the current implementation is possibly unsound: https://github.com/rust-lang/miri/issues/3780.

**Comment by RalfJung** [comment]
Turns out `join` is fine, the issue is deeper -- self-referential async block lowering is [still unsound](https://github.com/rust-lang/rust/issues/63818#issuecomment-2264915918).

**Comment by bonsairobo** [comment]
Any plans for a similar `try_join!` macro?
